### PR TITLE
VACMS-15566: Lovell Listing Pages (2)

### PR DESCRIPTION
**Pull request by ryguyk**
## Description
Closes https://github.com/department-of-veterans-affairs/va.gov-cms/issues/15566

## QA steps (WIP)
- Lovell bifurcated stories:
  - [x] Confirm Lovell bifurcated stories are built for both TRICARE and VA.
  - [x] Confirm that bifurcated story has (appropriate) switch link.
  - [x] Confirm that bifurcated story has link back to appropriate listing page.
    - E.g.
      - TRICARE: [/lovell-federal-health-care-tricare/stories/test-story-for-both-lovell/](https://pr209-kszltj1fmi0zxkwondkyrbq4pziwfyzl.tugboat.vfs.va.gov/lovell-federal-health-care-tricare/stories/test-story-for-both-lovell/)
      - VA: [/lovell-federal-health-care-va/stories/test-story-for-both-lovell/](https://pr209-kszltj1fmi0zxkwondkyrbq4pziwfyzl.tugboat.vfs.va.gov/lovell-federal-health-care-va/stories/test-story-for-both-lovell/)

- Lovell TRICARE-only stories:
  - [x] Confirm Lovell TRICARE stories do not have switch link.
  - [x] Confirm Lovell TRICARE stories link back to TRICARE listing page.
    - E.g.
      - [/lovell-federal-health-care-tricare/stories/test-story-for-tricare-lovell/](https://pr209-kszltj1fmi0zxkwondkyrbq4pziwfyzl.tugboat.vfs.va.gov/lovell-federal-health-care-tricare/stories/test-story-for-tricare-lovell/)

- Lovell VA-only stories
  - [x] Confirm Lovell VA stories do not have switch link.
  - [x] Confirm Lovell VA stories link back to VA listing page.
    - E.g.
      - [/lovell-federal-health-care-va/stories/test-story-for-va-lovell/](https://pr209-kszltj1fmi0zxkwondkyrbq4pziwfyzl.tugboat.vfs.va.gov/lovell-federal-health-care-va/stories/test-story-for-va-lovell/)

- Lovell TRICARE listing page:
    - [x] Confirm Lovell bifurcated story from above shows up on TRICARE listing page.
    - [x] Confirm Lovell TRICARE story from above shows up on TRICARE listing page.
    - [x] Confirm Lovell VA story from above does NOT show up on TRICARE listing page.
      - Listing Page:
        - TRICARE: [/lovell-federal-health-care-tricare/stories/](https://pr209-kszltj1fmi0zxkwondkyrbq4pziwfyzl.tugboat.vfs.va.gov/lovell-federal-health-care-tricare/stories/)

- Lovell VA listing page:
    - [x] Confirm Lovell bifurcated story from above shows up on VA listing page.
    - [x] Confirm Lovell VA story from above shows up on VA listing page.
    - [x] Confirm Lovell TRICARE story from above does NOT show up on VA listing page.
      - Listing Page:
        - VA: [/lovell-federal-health-care-va/stories/](https://pr209-kszltj1fmi0zxkwondkyrbq4pziwfyzl.tugboat.vfs.va.gov/lovell-federal-health-care-va/stories/)
       
- Other listing pages: 
  - [x] Confirm non-Lovell listing pages render appropriately with pager.
    - E.g. 
      - [x] [/minneapolis-health-care/stories](https://pr209-kszltj1fmi0zxkwondkyrbq4pziwfyzl.tugboat.vfs.va.gov/minneapolis-health-care/stories/) should have three (3) pages.
      - [x] [/eastern-oklahoma-health-care/stories](https://pr209-kszltj1fmi0zxkwondkyrbq4pziwfyzl.tugboat.vfs.va.gov/eastern-oklahoma-health-care/stories/) should have seven (7) pages.

## Screenshots
image

### CODE REVIEW

**Comment by ryguyk** [comment]
Opened a new PR to try to get as fresh an environment as possible [after fighting some things on Tugboat](https://github.com/department-of-veterans-affairs/next-build/pull/198). No difference.

File: src/lib/drupal/lovell/constants.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
+import { RESOURCE_TYPES } from '@/lib/constants/resourceTypes'
```
Comment (tjheffner) [inline src/lib/drupal/lovell/constants.ts:1]: big ++ to moving this stuff to its own `lovell/` dir

File: src/lib/drupal/lovell/utils.ts [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+  LovellChildVariant,
+  LovellFormattedResource,
+  LovellBifurcatedFormattedResource,
+} from './types'
+import {
+  LOVELL_RESOURCE_TYPES,
+  LOVELL_BIFURCATED_RESOURCE_TYPES,
+} from './constants'
+import { StaticPathResourceType } from '@/types/index'
+import { FormattedResource, QUERIES_MAP } from '@/data/queries'
```
[the remark this answers is not in the document]
Comment (tjheffner) [reply src/lib/drupal/lovell/utils.ts:13]: this appears valid in here. maybe you did/were going to use the same `(typeof QUERIES_MAP)[(typeof LOVELL_RESOURCE_TYPES)[number]]['formatter']` pattern from before but changed it here?

File: src/lib/drupal/lovell/staticProps.ts [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+  return {
+    ...resource,
+    entityPath: variantPaths[variant],
+    socialLinks: {
+      ...resource.socialLinks,
+      path: variantPaths[variant],
+    },
+    administration: LOVELL[variant].administration,
+    listing: getLovellVariantOfUrl(resource.listing, variant),
+    canonicalLink: variantPaths.va,
```
Comment (tjheffner) [inline src/lib/drupal/lovell/staticProps.ts:81]: 😎

File: src/lib/drupal/listingPages.ts [hunk 4]
```diff
@@ -25,62 +32,95 @@
...
     return []
   }
 
   return Promise.all(
     listingPageStaticPathResources.map(async (resource) => {
-      const totalPages = await getListingPageCount(
-        resource,
-        listingResourceType
-      )
+      const { totalItems: itemCount, totalPages: pageCount } =
```
[the remark this answers is not in the document]
Comment (tjheffner) [reply src/lib/drupal/listingPages.ts:91]: this also seems valid, with the change on line 115

**Comment by tjheffner** [comment]
I hit "submit review" before I had a chance to type this in the box... Overall this PR represents a mountain of work, really remarkable work. Thank you for tackling this so thoroughly. 👏 Only a few small comments and questions, noted above.

With tugboat's ongoing issues, I pulled locally & targeted my local ddev endpoint. This resulted in a successful build of 2961 pages (I haven't pulled a fresh db in a minute). Verified lovell listing pages had stories and could switch between them.  Verified non-lovell story & story listings function as intended.

```
yarn test
Test Suites: 47 passed, 47 total
Tests:       215 passed, 215 total
Snapshots:   13 passed, 13 total
Time:        3.465 s
Ran all test suites.
```

File: src/lib/constants/resourceTypes.ts [hunk 5]
```diff
@@ -6,3 +6,7 @@ export const RESOURCE_TYPES = {
 
 export type ResourceTypeType =
   (typeof RESOURCE_TYPES)[keyof typeof RESOURCE_TYPES]
+
+export const ADDITIONAL_RESOURCE_TYPES = {
```
[the remark this answers is not in the document]
Comment (ryguyk) [reply src/lib/constants/resourceTypes.ts:10]: I think we need some way of separating _Drupal_ resources from other things (e.g. `static-path-resources`) that are only keys into the query layer. This becomes clear when we look at `ResourceTypeType` (to be renamed soon), which is an enum representing the possible resource types that we get from fetching Drupal data. While we use `static-path-resources` to key into the query layer, we also pass in a `resourceType`, which is of type `ResourceTypeType`. We need a way to isolate the real Drupal resource types from "helper" resource types.

I'm open to discussing naming, etc. I'm not stubbornly committed to anything specific. But I do think we need to have something like this in order to keep them separate.

File: src/data/queries/staticPathResources.ts [hunk 6]
```diff
@@ -44,41 +47,13 @@ export const data: QueryData<
...
-
-  // 4. Glue all pages together.
-  return [firstPageData, ...subsequentPageData].flat()
+  return (
+    await fetchAndConcatAllResourceCollectionPages<JsonApiResourceWithPathAndFieldAdmin>(
+      opts.resourceType,
+      params(opts.resourceType),
+      PAGE_SIZE
+    )
+  ).data
```
[the remark this answers is not in the document]
Comment (ryguyk) [reply src/data/queries/staticPathResources.ts:56]: I changed this to use object destructuring. It's not different practically, but perhaps it seems more clear?

File: src/data/queries/index.ts [hunk 7]
```diff
@@ -43,4 +44,10 @@ export const queries = createQueries({
 
   // Static Path Generation
   'static-path-resources': StaticPathResources,
-})
+}
+
+export type FormattedResource = ReturnType<
+  (typeof QUERIES_MAP)[(typeof RESOURCE_TYPES)[keyof typeof RESOURCE_TYPES]]['formatter']
```
[the remark this answers is not in the document]
Comment (ryguyk) [reply src/data/queries/index.ts:50]: I added a comment here and for the similar definition for listing pages. It should help clarify what's going on.

[the remark this answers is not in the document]
Comment (tjheffner) [reply src/data/queries/index.ts:57]: thank you!

File: src/lib/constants/resourceTypes.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (tjheffner) [reply src/lib/constants/resourceTypes.ts:10]: gotcha, that makes sense to me. let's return to the naming scheme in a follow up, we can iterate

**Review by tjheffner — APPROVED** [review]

**Pull request merged**